### ISSUE TITLE
Make `StringViewArray::slice()` and `BinaryViewArray::slice()` faster

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#6408]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
While working on an upstream project https://github.com/apache/datafusion/pull/12092 which switched DataFusion to use `StringViewArray` rather than `StringArray`

When I did, one of the queries got much slower. 

Profiling revealed that the time difference was almost entirely explained by the time spent in `StringViewArray::slice()` 

![Screenshot 2024-09-16 at 4 44 50 PM](https://github.com/user-attachments/assets/efe90a30-1188-4af4-822c-b584b177770a)

Here is the flamegraph with `StringArray`:
![flamegraph-main](https://github.com/user-attachments/assets/00216a16-7688-4cbb-bc96-372841bb157c)

Here is the same query with `StringViewArray`:
![flamegraph-string-view](https://github.com/user-attachments/assets/1a314737-a0ba-4f00-8a6a-0a62626a01ce)

I am pretty sure the additional time is due to the time spent allocating / copying / deallocating the `Vec`s of buffers here:

https://github.com/apache/arrow-rs/blob/341ec357e74e897f50250930b44f453bce54a19a/arrow-array/src/array/byte_view_array.rs#L118

Where calling `slice` on a StringArray can be done with a few Arc increments. 

https://github.com/apache/arrow-rs/blob/3490639252294215c7ee05990d82b43e2cd097a6/arrow-array/src/array/byte_array.rs#L88-L91

**Describe the solution you'd like**
I would like `StringViewArray::slice` to be faster (aka don't allocate)

**Describe alternatives you've considered**

We can (and probably should) change DataFusion not to use slice in this case (I am working to file a ticket on this) but I think making `slice` faster / non allocating for `StringViewArray` will be useful in general

**Additional context**

**Comment by alamb** [comment]
fyi @a10y and @XiangpengHao

**Comment by alamb** [comment]
On solution here would be to replace `Vec<Buffer>` with `Arc<[Buffer]>` after construction -- that would make slice extremely cheap

**Comment by XiangpengHao** [comment]
> On solution here would be to replace Vec<Buffer> with Arc<[Buffer]> after construction 

I agree, this should fix the problems. In fact, many other operations (e.g., take) also need to clone the Vec<Buffer>, changing to Arc<[Buffer]> will benefit them as well.

However, I would be a bit more careful about why cloning the buffer taking so long -- often indicating the Vec<Buffer> is large, which often means gc is not being called timely. So in addition to changing to Arc<[Buffer]>, I would also examine the plan to make sure gc (in CoalesceBatchesExec) is being invoked properly. cc @WetABQ who might be interested in this discussion.

I plan to work on this later this week but anyone else feel free to run faster than me!

**Comment by XiangpengHao** [comment]
> replace Vec<Buffer> with Arc<[Buffer]> after construction -- that would make managing the buffers on slice extremely cheap

But that also means an extra indirection to read buffers, not sure if it matters

**Comment by Rachelint** [comment]
It may 

> On solution here would be to replace `Vec<Buffer>` with `Arc<[Buffer]>` after construction -- that would make managing the buffers on `slice` extremely cheap

The slice maybe can't be eliminated until the epic https://github.com/apache/datafusion/issues/7065 is finished...
Because it will still be called here https://github.com/apache/datafusion/blob/a08f923c2acb1a46614970231d9a672c36ce3ad2/datafusion/physical-plan/src/aggregates/row_hash.rs#L713

**Comment by alamb** [comment]
> But that also means an extra indirection to read buffers, not sure if it matters

I think an `Arc<Vec<Buffer>>` would add another indirection. I was thinking/hoping that `Arc<[Buffer]>` would only still have one indirection

**Comment by alamb** [comment]
> The slice maybe can't be eliminated until the epic https://github.com/apache/datafusion/issues/7065 is finished...

FWIW the slice that I looked at in  https://github.com/apache/datafusion/pull/12092#issuecomment-2354023193 is a different one: 

https://github.com/apache/datafusion/blob/a08f923c2acb1a46614970231d9a672c36ce3ad2/datafusion/functions-aggregate-common/src/aggregate/groups_accumulator.rs#L435-L438

(This is called once for each distinct group in each batch being aggregates, which is quite bad  -- the  better way to solve this is to implement a Min/Max accumulator for strings that avoids slicing at all, which we are tracking in https://github.com/apache/datafusion/issues/6906)

I think the fact that slice is used many different places makes it all the more important to optimize in arrow-rs

**Comment by alamb** [comment]
> However, I would be a bit more careful about why cloning the buffer taking so long -- often indicating the Vec is large, which often means gc is not being called timely

This is a good point

**Comment by Rachelint** [comment]
> > The slice maybe can't be eliminated until the epic [apache/datafusion#7065](https://github.com/apache/datafusion/issues/7065) is finished...
> 
> FWIW the slice that I looked at in [apache/datafusion#12092 (comment)](https://github.com/apache/datafusion/pull/12092#issuecomment-2354023193) is a different one:
> 
> https://github.com/apache/datafusion/blob/a08f923c2acb1a46614970231d9a672c36ce3ad2/datafusion/functions-aggregate-common/src/aggregate/groups_accumulator.rs#L435-L438
> 
> (This is called once for each distinct group in each batch being aggregates, which is quite bad -- the better way to solve this is to implement a Min/Max accumulator for strings that avoids slicing at all, which we are tracking in [apache/datafusion#6906](https://github.com/apache/datafusion/issues/6906))
> 
> I think the fact that slice is used many different places makes it all the more important to optimize in arrow-rs

Yes, maybe it should be ensured to be a cheap operations, it is used in many many places...

**Comment by ShiKaiWi** [comment]
@alamb @XiangpengHao Could assign this ticket to me? I guess I can make it.

BTW, I find there is no benchmark case for the `slice` method, and I guess the benchmark case is necessary to prove the proposal implementation.

**Comment by XiangpengHao** [comment]
You can self assign by commenting "take"

I think adding benchmark is the right first step! Looking forward to it!

**Comment by alamb** [comment]
Tahank you @ShiKaiWi  -- can't wait to check it out

**Comment by alamb** [comment]
This seems to have come up again in datafusion, see:
- https://github.com/apache/datafusion/pull/18161
